### PR TITLE
Add luminance-sdl2 backend.

### PR DESCRIPTION
**Pull request by iwikal**
I've been using luminance with sdl2 for a while now and I think it works great. I thought I'd share what I built, but I didn't realize you had this massive 0.40 rewrite on the way.
I rewrote the Sdl2Surface to my best understanding of the new architecture, but I haven't ported my game to 0.40 yet, so I haven't really tested this incarnation of the surface. I figured I might as well open the PR now, and mark it as ready once I've confirmed it still works.

### CODE REVIEW

**Comment by phaazon** [comment]
Woah, sick! Thanks a lot! I’ll review that as soon as possible, thanks a lot!

**Comment by iwikal** [comment]
After reading about your plans of deprecating luminance-windowing and looking closer at luminance-glutin I think it's the right choice for luminance-sdl2 too. Hiding all backend functionality behind a common abstracted API kind of defeats the purpose of having different backends to choose from, and providing both the abstract API and raw access feels messy.

**Comment by phaazon** [comment]
Yep, completely. That’s why I don’t re-export symbols from the backend crates anymore. :)

File: luminance-sdl2/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+[package]
+name = "luminance-sdl2"
+version = "0.1.0"
+license = "BSD-3-Clause"
+authors = [
+  "Dimitri Sabadie <dimitri.sabadie@gmail.com>",
+  "Joel Nordström <e.joel.nordstrom@gmail.com>"
+]
+description = "Sdl2 support for luminance"
```
Comment (phaazon) [inline luminance-sdl2/Cargo.toml:9]: Can we use `SDL2` is they’re using it in the `sdl2` crate, please? :)

File: luminance-sdl2/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  /// graphics state has already been acquired.
+  GraphicsStateError(StateQueryError),
+}
+
+impl fmt::Display for Sdl2SurfaceError {
+  fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+    match *self {
+      Sdl2SurfaceError::InitError(ref e) => {
+        f.write_str("initialization error: ")?;
+        e.fmt(f)
```
Comment (phaazon) [inline luminance-sdl2/src/lib.rs:41]: You can simplify with

```suggestion
        write!(f, "initialization error: {}", e)
```

File: luminance-sdl2/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+      }
+      Sdl2SurfaceError::WindowCreationFailed(ref e) => {
+        f.write_str("failed to create window: ")?;
+        e.fmt(f)
```
Comment (phaazon) [inline luminance-sdl2/src/lib.rs:45]: Ditto.

File: luminance-sdl2/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+      }
+      Sdl2SurfaceError::GlContextInitFailed(ref e) => {
+        f.write_str("failed to create OpenGL context: ")?;
+        e.fmt(f)
```
Comment (phaazon) [inline luminance-sdl2/src/lib.rs:49]: Ditto.

File: luminance-sdl2/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+      }
+      Sdl2SurfaceError::VideoInitError(ref e) => {
+        f.write_str("failed to initialize video system: ")?;
+        e.fmt(f)
```
Comment (phaazon) [inline luminance-sdl2/src/lib.rs:53]: Ditto.

File: luminance-sdl2/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+      }
+      Sdl2SurfaceError::GraphicsStateError(ref e) => {
+        f.write_str("failed to get graphics state: ")?;
+        e.fmt(f)
```
Comment (phaazon) [inline luminance-sdl2/src/lib.rs:57]: Ditto.

File: luminance-sdl2/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+      }
+    }
+  }
+}
+
+/// A [luminance] GraphicsContext backed by Sdl2 and OpenGL 3.3 Core.
```
Comment (phaazon) [inline luminance-sdl2/src/lib.rs:63]: SDL2*

File: luminance-sdl2/src/lib.rs (hunk 2 above)
Comment (iwikal) [reply luminance-sdl2/src/lib.rs:41]: Yea, I don't know what I was thinking when I decided to do this microoptimization. It's probably not even faster, and even if it was, it's not worth the uglier code. I'll change it.

File: luminance-sdl2/Cargo.toml (hunk 1 above)
Comment (iwikal) [reply luminance-sdl2/Cargo.toml:9]: Absolutely!

**Comment by iwikal** [comment]
When I made this, I used `luminance-glfw` as a starting point and modified it to use sdl2 instead, but I think the `luminance-glutin` API is a great idea and I will gladly rewrite this in a similar way.

**Comment by iwikal** [comment]
After removing anything related to `luminance-windowing` and adding `GL33Surface::build_with` there wasn't much use to keeping `GL33Surface::new` around. It's a shame that the WindowBuilder chains with `&mut Self` instead of `Self` though, that forces the consumer code to be less functional and cool.

I'm not sure how to add SDL2 to the CI environment, is that even something I can do?

**Comment by phaazon** [comment]
It’s tested by default for everything listed as members in the front `Cargo.toml. You have to modify the `.github/workflows/ci.yml` file to add support for the `quality` job though — not required right now, I’ll do it if you don’t want to bother. :)

**Comment by iwikal** [comment]
Yea, if you don't mind I think it would be most convenient for you to do that. But I've successfully ported my little game project thing to use this, and it works like a charm!

**Comment by iwikal** [comment]
@phaazon It builds!

**Pull request merged**